lib/DependencyScan/ScanDependencies.cpp

```diff
@@ -1652,3 +1672,101 @@ swift::dependencies::performBatchModuleScan(
+    // already known to be out of date.
+    hasOutOfDateModuleDependency |=
+        (modulesRequiringRescan.find(depID) != modulesRequiringRescan.end());
+  }
+
+  if (hasOutOfDateModuleDependency) {
+    if (emitRemarks)
+      diags.diagnose(SourceLoc(), diag::remark_scanner_invalidate_upstream,
+                     moduleID.ModuleName);
+    modulesRequiringRescan.insert(moduleID);
+  } else if (!verifyModuleDependencyUpToDate(moduleID, cache, cacheTimeStamp,
+                                             fs, diags, emitRemarks))
+    modulesRequiringRescan.insert(moduleID);
+
+  visited.insert(moduleID);
+}
+
+bool swift::dependencies::incremental::verifyModuleDependencyUpToDate(
+    const ModuleDependencyID &moduleID, const ModuleDependenciesCache &cache,
+    const llvm::sys::TimePoint<> &cacheTimeStamp, llvm::vfs::FileSystem &fs,
+    DiagnosticEngine &diags, bool emitRemarks) {
+  const auto &moduleInfo = cache.findKnownDependency(moduleID);
+  auto verifyInputOlderThanCacheTimeStamp = [&cacheTimeStamp, &fs, &diags,
+                                             emitRemarks](StringRef moduleName,
+                                                          StringRef inputPath) {
+    llvm::sys::TimePoint<> inputModTime = llvm::sys::TimePoint<>::max();
+    if (auto Status = fs.status(inputPath))
+      inputModTime = Status->getLastModificationTime();
+    if (inputModTime > cacheTimeStamp) {
+      if (emitRemarks)
+        diags.diagnose(SourceLoc(),
+                       diag::remark_scanner_stale_result_invalidate, moduleName,
+                       inputPath);
+      return false;
+    }
+    return true;
+  };
+
+  // Check interface file for Swift textual modules
+  if (const auto &textualModuleDetails = moduleInfo.getAsSwiftInterfaceModule())
+    if (!verifyInputOlderThanCacheTimeStamp(
+            moduleID.ModuleName, textualModuleDetails->swiftInterfaceFile))
+      return false;
+
+  // Check binary module file for Swift binary-only modules
+  if (const auto &binaryModuleDetails = moduleInfo.getAsSwiftBinaryModule())
+    if (!verifyInputOlderThanCacheTimeStamp(
+            moduleID.ModuleName, binaryModuleDetails->compiledModulePath))
+      return false;
+
+  // Check header input source files (bridging header etc.)
+  for (const auto &headerInput : moduleInfo.getHeaderInputSourceFiles())
+    if (!verifyInputOlderThanCacheTimeStamp(moduleID.ModuleName, headerInput))
+      return false;
+
+  // Auxiliary files
+  for (const auto &auxInput : moduleInfo.getAuxiliaryFiles())
+    if (!verifyInputOlderThanCacheTimeStamp(moduleID.ModuleName, auxInput))
+      return false;
+
+  // Check header/modulemap source files for a Clang dependency
+  if (const auto &clangModuleDetails = moduleInfo.getAsClangModule())
+    for (const auto &fileInput : clangModuleDetails->fileDependencies)
+      if (!verifyInputOlderThanCacheTimeStamp(moduleID.ModuleName, fileInput))
+        return false;
+
```

Comment:
Can you add the verification for CASIDs in the dependencies? Just make a query to make sure the root node can be found is probably enough.

clang dependencies has CASFileSystemRootID, CASClangIncludeTreeRootID
swift textual has: CASFileSystemRootID, CASBridgingHeaderIncludeTreeRootID
binary modules are probably fine since the CASID is setup by the dependency module.

Reply:
After second thoughts, for better validation, maybe binary module should set up its own CASID and stored in the cache. Feel free to file an issue on me to add CAS validation to this module cache reuse.

For this PR, you can maybe leave the cache reuse to be off for caching build.

Reply:
Disabled incremental cache re-use for CAS builds, for now. 👍🏼